### PR TITLE
IN-799 Add VSCode / TypeScript template

### PR DESCRIPTION
**Pull request by borodean**
Introduces a basic VSCode / TypeScript template.

**Features:**

- Clones a remote repository to use as the TypeScript application source.
- Runs `npm test` to determine whether the challenge is solved.

**What's inside:**

The whole thing is built on top of the [`codercom/code-server`](https://hub.docker.com/r/codercom/code-server) container - that's the one that runs the VSCode server.

During the setup phase, it installs Node.js to the container, clones [the repository containing the application](https://github.com/borodean/typescript-example-app) we'll be running our tests against, and installs its dependencies. It also corrupts one of the application files to make the tests fail (it felt more appropriate to keep the application example in the remote repository fully functional).

During the check phase, it runs `npm tests` and exits with a success code if the tests pass.

During the solve phase, it checks out the working version of the corrupted file from the git tree.

There is not cleanup to be done.

**Possible TODOs:**

- The repository of the example application must be moved under the Instruqt organization (https://github.com/borodean/typescript-example-app).
- The image used as the track logo must be moved under the Instruqt CDN (https://cdn.jsdelivr.net/gh/remojansen/logo.ts@1ff5efb/ts.png).

### CODE REVIEW

File: vscode-typescript/config.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+version: "2"
+containers:
+- name: container
+  image: codercom/code-server
+  cmd: code-server workspace /app --auth none --disable-telemetry
```
Comment (matthisk) [inline vscode-typescript/config.yml:5]: Could we make the code-server operate in dark mode by default?

You could use the `--user-data-dir` flag for this and pass a config file at `user-data/User/settings.json`

```suggestion
  cmd: code-server workspace /app --user-data-dir user-data --auth none --disable-telemetry
```

Example configuration:

```
{
    "workbench.colorTheme": "Default Dark+",
    "workbench.statusBar.visible": false,
    "workbench.tips.enabled": false,
    "workbench.startupEditor": "newUntitledFile",
    "update.showReleaseNotes": false,
    "telemetry.enableTelemetry": false,
    "window.menuBarVisibility": "hidden",
    "workbench.activityBar.visible": false,
    "breadcrumbs.enabled": false,
    "breadcrumbs.filePath": "off",
    "workbench.editor.enablePreview": false,
    "workbench.editor.enablePreviewFromQuickOpen": false,
    "workbench.editor.highlightModifiedTabs": true,
    "workbench.editor.showTabs": false,
    "files.enableTrash": false,
    "workbench.enableExperiments": false,
    "editor.codeLens": false,
    "editor.colorDecorators": false,
    "editor.dragAndDrop": false,
    "editor.smoothScrolling": true,
    "editor.minimap.enabled": false,
    "editor.fontSize": 13,
    "workbench.colorCustomizations": {
        "sideBar.background": "#252F42",
        "sideBar.foreground": "#fff",
        "editor.background": "#1C2639",
        "list.activeSelectionBackground": "#344156",
        "list.focusBackground": "#344156",
        "list.hoverBackground": "#13171e",
        "list.inactiveSelectionBackground": "#344156"
    }
}
```

File: vscode-typescript/01-passing-tests/setup-container [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+echo "Running challenge setup script on host container"
+
+# Install Node.js
+# https://github.com/nodesource/distributions/blob/master/README.md
+curl -fsSL https://deb.nodesource.com/setup_14.x | bash -
+apt-get install -y nodejs
+
+# Clone the example repository
+git clone https://github.com/borodean/typescript-example-app.git /app
```
Comment (matthisk) [inline vscode-typescript/01-passing-tests/setup-container:10]: I have forked the repo under the Instruqt org: https://github.com/instruqt/typescript-example-app

File: vscode-typescript/config.yml (hunk 1 above)
Comment (borodean) [reply vscode-typescript/config.yml]: Updated: https://github.com/instruqt/templates/pull/3/commits/0e935cb74ae4be4de46c24fd8557eca5916217f4.

File: vscode-typescript/01-passing-tests/setup-container (hunk 2 above)
Comment (matthisk) [reply vscode-typescript/01-passing-tests/setup-container:18]: @borodean Can you update this URL to the fork on the instruqt org?

File: vscode-typescript/track.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  The track checks are failing based on the test results.
+
+  ## Using this template
+  To create your own track from this template, run this command:
+  ```
+  instruqt track create \
+      --from templates/container \
+      --title "My VSCode / TypeScript track"
+  ```
+icon: https://cdn.jsdelivr.net/gh/remojansen/logo.ts@1ff5efb/ts.png
```
Comment (matthisk) [inline vscode-typescript/track.yml:17]: I have uploaded this image to our CDN: https://cdn.instruqt.com/assets/templates/ts.png

File: vscode-typescript/track.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
+slug: vscode-typescript
+id: 0dxngfkws4op
+type: track
+title: VSCode / TypeScript
```
Comment (matthisk) [inline vscode-typescript/track.yml:4]: Quick suggestion:
```suggestion
title: VSCode & TypeScript
```

Comment (borodean) [reply vscode-typescript/track.yml]: Updated: https://github.com/instruqt/templates/pull/3/commits/de42e00e9b4058338a86ded2d6f89a1ad6820603.

File: vscode-typescript/track.yml (hunk 3 above)
Comment (borodean) [reply vscode-typescript/track.yml]: Updated: https://github.com/instruqt/templates/pull/3/commits/a3ae087085fc272048cfcbc985c12d8b719b9513.

File: vscode-typescript/01-passing-tests/setup-container (hunk 2 above)
Comment (borodean) [reply vscode-typescript/01-passing-tests/setup-container]: Updated: https://github.com/instruqt/templates/pull/3/commits/642f75d1c098aebd3a8f35e9dc6ac9a88ec92b4c.

**Review by matthisk — APPROVED** [review]

File: vscode-typescript/track.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+teaser: Edit and test TypeScript applications in VSCode
+description: |-
+  A basic track for editing a TypeScript application and running tests for it.
+  The track checks are failing based on the test results.
+
+  ## Using this template
+  To create your own track from this template, run this command:
+  ```
+  instruqt track create \
+      --from templates/container \
```
Comment (wietsevenema) [inline vscode-typescript/track.yml:14]: I believe this should be `templates/vscode-typescript`

File: vscode-typescript/track.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
+slug: vscode-typescript
+id: 0dxngfkws4op
+type: track
+title: VSCode & TypeScript
```
Comment (wietsevenema) [inline vscode-typescript/track.yml:4]: I believe this template exists to demonstrate VSCode. Consider to make that clear in the title and remove Typescript - it's a distractor.

File: vscode-typescript/track.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+teaser: Edit and test TypeScript applications in VSCode
+description: |-
+  A basic track for editing a TypeScript application and running tests for it.
```
Comment (wietsevenema) [inline vscode-typescript/track.yml:7]: Same comment - this is a VSCode template, not a TypeScript template

File: vscode-typescript/track.yml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  ```
+  instruqt track create \
+      --from templates/container \
+      --title "My VSCode & TypeScript track"
+  ```
+icon: https://cdn.instruqt.com/assets/templates/ts.png
+tags: []
+owner: templates
+developers:
+- borodean@gmail.com
```
Comment (wietsevenema) [inline vscode-typescript/track.yml:21]: Can you change this to the Instruqt account.

File: vscode-typescript/track.yml (hunk 6 above)
Comment (matthisk) [reply vscode-typescript/track.yml:4]: Copy is not final yet.

File: vscode-typescript/track.yml (hunk 8 above)
Comment (matthisk) [reply vscode-typescript/track.yml:21]: Yes this is probably a good idea. Not sure if we have an instruqt account, we should also make this consistent on the other templates.

Comment (borodean) [reply vscode-typescript/track.yml:21]: Updated in https://github.com/instruqt/templates/pull/4.

File: vscode-typescript/track.yml (hunk 5 above)
Comment (borodean) [reply vscode-typescript/track.yml:14]: Updated in https://github.com/instruqt/templates/pull/4.

**Pull request merged**